libsupport/include/katana/Cache.h

```diff
@@ -0,0 +1,157 @@
+#ifndef KATANA_LIBSUPPORT_KATANA_CACHE_H_
+#define KATANA_LIBSUPPORT_KATANA_CACHE_H_
+
+// This is single threaded only, it is not intended to store large objects,
+// but rather metadata (e.g., a shared_ptr to a property column).
+
+// The problem witchel had implementing a multi-threaded version using
+// parallel-hashmap is a lock ordering problem.  parallel-hashmap 1.33 allows
+// execution of code with the write lock held, only when modifying, not when adding an
+// element.  We have to modify the LRU list when eviciting an element, so the natural
+// lock ordering is parallel-hashmap write lock, then list lock.  But without a way to
+// execute insert code with the parallel-hashmap write lock held, it seemed like there
+// would be some form of race condition.
+
+#include <list>
+#include <optional>
+#include <string>
+
+#include <boost/container_hash/hash.hpp>
+
+#include "katana/ArrowInterchange.h"
+#include "katana/Logging.h"
+
+namespace tsuba {
+
+template <typename Key, typename Value>
+class KATANA_EXPORT Cache {
+  using ListType = std::list<Key>;
+  struct MapValue {
+    Value value;
+    // This allows us to delete the old position in the LRU list without a scan
+    typename ListType::iterator lru_it;
+  };
+  using MapType = std::unordered_map<Key, MapValue>;
+
+public:
+  enum class ReplacementPolicy { kLRUSize, kLRUBytes };
+
+  Cache(
```

Comment:
It is somewhat awkward to support all arguments when they are only valid for some cases.

Either:
- Create separate constructors for the Vanilla LRU and the Bytes LRU, e.g., `Cache(capacity, evict_cb)` and `Cache(capacity, evict_cb, value_to_bytes)` as the bytes policy requires `value_to_bytes`
- Create separate factory methods `template<typename Key, typename Value> static Cache<Key,Value> MakeFoo(...)`

Reply:
I have done option 1.

Reply:
Oh, I have not.  Hold on!

Reply:
Ok, sorry, good to go.